### PR TITLE
Tests

### PR DESCRIPTION
**Pull request by paquiteau**
This PR add unit test (#4) for the nufft operators. 

- [X] Basic tests on NFFT == NDFT (and special FFT case) 
- [X] Adjoint property
- [X] Accuracy of forward (`op`) and adjoint (`adj_op`), compared to NFFT
- [X] Batch computation (Cufinufft & finufft only) 
- [X] Density Compensation (check that the fixed point iteration converge), cufinufft only
- [ ] Initializaion of data on device (cufinufft only)
- [x] Data Consistency functions

Only a subset of trajectory are used to initialized the operators (see the `case_trajectory.py` file)

the pynufft and tensorflow backend are not tested (yet)

### CODE REVIEW

**Comment by chaithyagr** [comment]
Ill review this and then merge, we can work on the GPU pipelines as a different PR maybe

File: tests/case_trajectories.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        return freq_2d.reshape(-1, 2), (N, N)
+
+    def case_grid3D(self, N=16):
+        """Create a 3D cartesian grid of frequencies locations."""
+        freq_1d = sp.fft.fftfreq(N)
+        freq_3d = np.stack(np.meshgrid(freq_1d, freq_1d, freq_1d), axis=-1)
+        return freq_3d.reshape(-1, 3), (N, N, N)
+
+
+def case_grid1D(N=256):
```
Comment (chaithyagr) [inline tests/case_trajectories.py:53]: Maybe a typo on the indexing? Why is this not in `CaseTrajectories` ?

File: tests/test_batch.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+from mrinufft.operators.interfaces import CUFINUFFT_AVAILABLE
+from case_trajectories import CasesTrajectories
+
+
+@fixture(scope="module")
+@parametrize(
+    backend=[
+        "finufft",
+        pytest.param(
+            "cufinufft",
```
Comment (chaithyagr) [inline tests/test_batch.py:20]: I recommend this as a separate test, with a tag like `def test_gpu_<name>` 
That way, we can make a CI just to run this. 
Or we could also add an environment variable `ONLY_RUN_GPU` which can be enabled to skip all tests except the ones needing GPU. WDYT? 

That way the CPU tests can happen online while GPU ones can happen offline, giving faster CI run times

File: tests/test_density.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    cases=[CasesTrajectories.case_radial3D],
+)
+@pytest.mark.skipif(not CUFINUFFT_AVAILABLE, reason="cufinufft not yet implemented")
+def test_density_pipe(kspace_traj, shape):
+    """Test the density compensation estimations."""
+    density = pipe(kspace_traj, shape, num_iter=20, tol=2e-7).get()
+    density_ref = pipe(kspace_traj, shape, num_iter=25, tol=2e-7).get()
+
+    # TODO: get tighter bounds.
+    npt.assert_allclose(density, density_ref, atol=1, rtol=1)
```
Comment (chaithyagr) [inline tests/test_density.py:22]: A test of the density of `tensorflow-nufft` compared to `cufinufft` is also a great test here to ensure sanity. Maybe add that ? (I can handle in my PR if it is too much, though its just one function call: https://mrphys.github.io/tensorflow-mri/api_docs/tfmri/sampling/estimate_density/

File: tests/test_interfaces.py [hunk 4]
```diff
@@ -1,79 +1,102 @@
...
+@fixture(scope="module")
+@parametrize(
+    "backend",
+    [
+        "pynfft",
+        "finufft",
+        pytest.param(
+            "cufinufft",
+            marks=pytest.mark.skipif(
+                not CUFINUFFT_AVAILABLE, reason="cufinufft not yet implemented"
```
Comment (chaithyagr) [inline tests/test_interfaces.py:28]: Same comment as above

File: tests/test_interfaces.py [hunk 5]
```diff
@@ -1,79 +1,102 @@
...
-    fft_mat = fft_matrix[0]
-    if len(shape) == 2:
-        fft_mat = fft_matrix[0].flatten()[:, None] @ fft_matrix[1].flatten()[None, :]
-        fft_mat = (
-            fft_mat.reshape(shape * 2)
-            .transpose(2, 0, 1, 3)
-            .reshape((np.prod(shape),) * 2)
-        )
-    assert np.allclose(ndft_op._fourier_matrix, fft_mat)
+def test_interfaces_accuracy_backward(operator, kspace_data, nfft_ref_op):
```
Comment (chaithyagr) [inline tests/test_interfaces.py:87]: I dont understand how are these functions called for test? Am I missing some API for parameterize?

**Review by chaithyagr — CHANGES_REQUESTED** [review]
A bunch of minor changes and issues. Can you please comment or address them and add them back to the review?

Thank you for this PR, pretty well written code otherwise!

File: src/mrinufft/operators/interfaces/base.py [hunk 6]
```diff
@@ -103,10 +110,21 @@ def adj_op(self, coeffs):
         """
         pass
 
+    def with_off_resonnance_correction(self, B, C, indices):
```
[the remark this answers is not in the document]
Comment (paquiteau) [reply src/mrinufft/operators/interfaces/base.py:113]: It can be both. But I like the natural formalism of getting the augmented operator from the vanilla one. Anyway off-resonnance is half supported for now, so thats a debate that can be post-poned ;)

[the remark this answers is not in the document]
Comment (chaithyagr) [reply src/mrinufft/operators/interfaces/base.py:113]: Haha cool, resolving

File: src/mrinufft/operators/interfaces/base.py [hunk 7]
```diff
@@ -119,24 +137,117 @@ def shape(self, shape):
...
+
+    def __repr__(self):
+        """Return info about the Fourier operator."""
+        return (
+            f"{self.__class__.__name__}(\n"
+            f"  shape: {self.shape}\n"
+            f"  n_coils: {self.n_coils}\n"
+            f"  n_samples: {self.n_samples}\n"
+            f"  uses_sense: {self._uses_sense}\n"
+            ")"
```
[the remark this answers is not in the document]
Comment (paquiteau) [reply src/mrinufft/operators/interfaces/base.py:226]: Thats going to show the full array, I would rather display `self.uses_density`.

File: src/mrinufft/operators/interfaces/cufinufft.py [hunk 8]
```diff
@@ -78,16 +137,27 @@ def _make_plan(self, typ, **kwargs):
...
+        self.plans[typ]._references = [x, y]
+
         fpts_axes = [get_ptr(y), get_ptr(x), None]
         if self.ndim == 3:
-            z = cp.ascontiguousarray(self.samples[:, 2])
+            z = cp.array(self.samples[:, 2])
+            self.plans[typ]._references.append(z)
             fpts_axes.insert(0, get_ptr(z))
         M = x.size
+        print(M)
```
[the remark this answers is not in the document]
Comment (paquiteau) [reply src/mrinufft/operators/interfaces/cufinufft.py:150]: yup

[the remark this answers is not in the document]
Comment (chaithyagr) [reply src/mrinufft/operators/interfaces/cufinufft.py:150]:
```suggestion
```

File: src/mrinufft/operators/interfaces/cufinufft.py [hunk 9]
```diff
@@ -147,14 +217,17 @@ class MRICufiNUFFT(FourierOperatorBase):
     Cufinufft is able to run multiple transform in parallel, this is controlled
     by the n_trans parameter. The data provided should be of shape, (n_batch,
     n_coils, img_shape) for op (type2) and (n_batch, n_coils, n_samples) for
-    adjoint (type1). and in contiguous memory order. For now only single
-    precision (float32 and complex64) is supported
+    adjoint (type1). and in contiguous memory order.
+
+    For now only single precision (float32 and complex64) is supported
```
[the remark this answers is not in the document]
Comment (paquiteau) [reply src/mrinufft/operators/interfaces/cufinufft.py:222]: This was a constraint with the previous bindings, actually I think we should support both.

File: src/mrinufft/operators/interfaces/cufinufft.py [hunk 10]
```diff
@@ -673,7 +770,76 @@ def __repr__(self):
...
+        n_samples,
+        *fpts_axes,
+        get_ptr(c),
+        get_ptr(f),
+        opts,
+        tol,
+    )
+
+
+def pipe(kspace, grid_shape, num_iter=10, tol=2e-7):
```
[the remark this answers is not in the document]
Comment (paquiteau) [reply src/mrinufft/operators/interfaces/cufinufft.py:810]: I disagree. The pipe method is implementation dependent, so it should reside as close as possible to the backend interface it uses. If in the end we have a pipe method for every interfaces, we could later have a `pipe_density(kspace, grid, num_iter, backend)` in the `trajectories/utils/density.py`. 

Also, having the pipe method here avoid a circular import.

File: tests/case_trajectories.py (hunk 1 above)
Comment (paquiteau) [reply tests/case_trajectories.py:53]: Because this is such a special case only use for the NDFT <-> FFT comparison. Having it in the class would create too much edge cases otherwise (not all nufft backend support 1D nufft)

File: tests/test_batch.py (hunk 2 above)
Comment (paquiteau) [reply tests/test_batch.py:20]: instead of rewriting the test for gpu cases (that would create a lot of redundancy) we can use pytest custom markers[^1] and command line argument. 

[^1]: https://docs.pytest.org/en/7.1.x/example/markers.html

File: tests/test_density.py (hunk 3 above)
Comment (paquiteau) [reply tests/test_density.py:22]: Tensorflow-nufft is not yet tested (I have to install it properly) but then yes that could be a good sanity check.

File: tests/test_interfaces.py (hunk 5 above)
Comment (paquiteau) [reply tests/test_interfaces.py:87]: see https://smarie.github.io/python-pytest-cases/ for the details.
TLDR: the tuple (kspace_grid, shape) is generate by each case function selected via `cases` argument. This is a simpler interface to pytest's fixtures

File: src/mrinufft/operators/interfaces/base.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (chaithyagr) [reply src/mrinufft/operators/interfaces/base.py:226]: Yes sure, I was looking for it and didn't find in review tab xD.. Resolving..

File: tests/test_density.py (hunk 3 above)
Comment (chaithyagr) [reply tests/test_density.py:20]: Ill see what to do in next PR

File: tests/case_trajectories.py (hunk 1 above)
Comment (chaithyagr) [reply tests/case_trajectories.py:53]: Ahhh i see cool

Comment (chaithyagr) [reply tests/case_trajectories.py:53]: Maybe add this point as a comment above?

File: tests/test_interfaces.py (hunk 5 above)
Comment (paquiteau) [reply tests/test_interfaces.py:81]: the operator argument is auto-magically detected as a fixture by pytest (because there is function with the same name and a fixture decorator). All the parametrization is done on the operator, and then each test is run with it.

Comment (chaithyagr) [reply tests/test_interfaces.py:81]: Ahhh cool, yes got it!

File: src/mrinufft/operators/interfaces/cufinufft.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (chaithyagr) [reply src/mrinufft/operators/interfaces/cufinufft.py:810]: I agrewe with respect to the circular import!

File: src/mrinufft/operators/interfaces/cufinufft.py [hunk 11]
```diff
@@ -183,65 +257,61 @@ def __init__(
...
+        self.samples = np.asfortranarray(
+            proper_trajectory(samples, normalize="pi").astype(np.float32)
+        )
+        self.dtype = self.samples.dtype
 
         # density compensation support
         if density is True:
-            self.density_d = MRICufiNUFFT.estimate_density(samples_d, shape)
-            self.uses_density = True
+            # TODO estimate the density using pipe method.
```
[the remark this answers is not in the document]
Comment (chaithyagr) [reply src/mrinufft/operators/interfaces/cufinufft.py:269]: Maybe you missed this?

File: src/mrinufft/operators/interfaces/cufinufft.py [hunk 12]
```diff
@@ -469,21 +563,21 @@ def _adj_op_sense(self, coeffs, img_d=None):
     def _adj_op_calibless(self, coeffs, img_d=None):
         coeffs_f = coeffs.flatten()
         n_trans_samples = self.n_trans * self.n_samples
-        ksp_batched = cp.empty(n_trans_samples, dtype=np.complex64)
+        ksp_batched = cp.empty(n_trans_samples, dtype=self.cpx_dtype)
         if self.uses_density:
             density_batched = cp.repeat(
                 self.density_d[None, :], self.n_trans, axis=0
```
Comment (chaithyagr) [inline src/mrinufft/operators/interfaces/cufinufft.py:569]:
```suggestion
                self.density[None, :], self.n_trans, axis=0
```
